src/test/shell/bazel/cc_integration_test.sh

```diff
@@ -2117,4 +2117,136 @@ EOF
   expect_log "Hello from main.cpp"
 }
 
+function test_cpp20_modules_with_clang() {
+  type -P clang || return 0
+  # Check if clang version is less than 17
+  clang_version=$(clang --version | head -n1 | grep -oE '[0-9]+\.[0-9]+' | head -n1)
+  if [[ -n "$clang_version" ]]; then
+    major_version=$(echo "$clang_version" | cut -d. -f1)
+    if [[ "$major_version" -lt 17 ]]; then
+      return 0
+    fi
+  fi
+
+  add_rules_cc "MODULE.bazel"
+  add_to_bazelrc "common --repo_env=CC=clang"
+  # TODO: Make it so that --cxxopt applies to module_interfaces as well.
+  add_to_bazelrc "common --copt=-std=c++20"
+
+  cat > BUILD.bazel <<'EOF'
+load("@rules_cc//cc:defs.bzl", "cc_library", "cc_binary")
+
+package(features = ["cpp_modules"])
+
+cc_library(
+  name = "base",
+  module_interfaces = ["base.cppm"],
+)
+cc_library(
+  name = "foo",
+  module_interfaces = ["foo.cppm"],
+  deps = [":base"]
+)
+cc_library(
+  name = "bar",
+  module_interfaces = ["bar.cppm"],
+  deps = [":base"]
+)
+cc_binary(
+  name = "main",
+  srcs = ["main.cc"],
+  deps = [":foo", ":bar"]
+)
+EOF
+  cat > main.cc <<'EOF'
+import foo;
+import bar;
+
+void f() {
+  f_foo();
+  f_bar();
+}
+
+int main() {
+  f();
+  return 0;
+}
+EOF
+  cat > foo.cppm <<'EOF'
+export module foo;
+import base;
+
+export void f_foo() {
+  f_base();
+}
+EOF
+  cat > bar.cppm <<'EOF'
+export module bar;
+import base;
+
+export void f_bar() {
+  f_base();
+}
+EOF
+  cat > base.cppm <<'EOF'
+export module base;
+
+export void f_base() {
+}
+EOF
+
+  bazel build //:main --experimental_cpp_modules --disk_cache=disk || fail "Expected build C++20 Modules success with compiler 'clang'"
+
+  # Verify that the build can hit the cache without action cycles.
+  bazel clean || fail "Expected clean success"
+  bazel build //:main --experimental_cpp_modules --disk_cache=disk || fail "Expected build C++20 Modules success with compiler 'clang'"
```

Comment:
```suggestion
  bazel build //:main --experimental_cpp_modules --disk_cache=disk || fail "Expected build C++20 Modules success with compiler 'clang'"
  expect_log "20 disk cache hit"
```

adds some additional assurance.

I have an idea how to solve this, but it doesn't resolve the other cycle. I'll keep looking.